### PR TITLE
feat: generate and verify signatures

### PR DESCRIPTION
**Pull request by nonvis**

### CODE REVIEW

File: src/noir/consensus/priv_validator.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#include <noir/codec/scale.h>
+#include <noir/consensus/priv_validator.h>
+
+namespace noir::consensus {
+
+std::optional<std::string> priv_validator::sign_vote(vote& vote_) {
+  // TODO: add some validation checks
+
+  // auto bz = codec::scale::encode(vote_);
+  auto bz = codec::scale::encode(static_cast<p2p::vote_message>(vote_));
```
Comment (xedonman) [inline src/noir/consensus/priv_validator.cpp]: adding below line to <vote.h> would disacard static_cast
NOIR_FOR_EACH_FIELD_DERIVED(noir::consensus::vote, noir::p2p::vote_message)
please note that <scale.h> should be under <vote.h>

File: src/noir/consensus/priv_validator.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  // TODO: save_signed; why?
+  vote_.signature = sig;
+  return {};
+}
+
+std::optional<std::string> priv_validator::sign_proposal(proposal& proposal_) {
+  // TODO: add some validation checks
+
+  // auto bz = codec::scale::encode(proposal_);
+  auto bz = codec::scale::encode(static_cast<p2p::proposal_message>(proposal_));
```
Comment (xedonman) [inline src/noir/consensus/priv_validator.cpp]: adding below line to <types.h> would disacard static_cast
NOIR_FOR_EACH_FIELD_DERIVED(noir::consensus::proposal, p2p::proposal_message)
please note that <scale.h> should be under <types.h>

File: src/noir/consensus/vote.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+  // Check signature
+  if (val->pub_key_.address() != val_addr) {
+    elog("add_vote() failed: invalid validator address");
+    return false;
+  }
+  p2p::vote_message temp(vote_.value());
+  temp.signature = {}; // TODO: debug why manual reset is required for serialization
+  // auto data_vote = codec::scale::encode(static_cast<p2p::vote_message>(vote_.value()));
+  auto data_vote = codec::scale::encode(temp);
```
Comment (xedonman) [inline src/noir/consensus/vote.cpp]: adding below line to <vote.h> would disacard static_cast
NOIR_FOR_EACH_FIELD_DERIVED(noir::consensus::vote, noir::p2p::vote_message)
please note that <scale.h> should be under <vote.h>

File: src/noir/consensus/priv_validator.cpp (hunk 1 above)
Comment (nonvis) [reply src/noir/consensus/priv_validator.cpp]: got it

File: src/noir/consensus/priv_validator.cpp (hunk 2 above)
Comment (nonvis) [reply src/noir/consensus/priv_validator.cpp]: got it

File: src/noir/consensus/vote.cpp (hunk 3 above)
Comment (nonvis) [reply src/noir/consensus/vote.cpp]: got it

**Review by xedonman — APPROVED** [review]

**Pull request merged**